fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerRegistryImpl.java

```diff
@@ -0,0 +1,130 @@
+ */
+public class LuceneAnalyzerRegistryImpl implements LuceneAnalyzerRegistry {
+    @Nonnull
+    private static final Logger LOGGER = LoggerFactory.getLogger(LuceneAnalyzerRegistryImpl.class);
+    @Nonnull
+    private static final LuceneAnalyzerRegistryImpl INSTANCE = new LuceneAnalyzerRegistryImpl();
+
+    @Nonnull
+    private Map<String, LuceneAnalyzerFactory> registry;
+
+    @Nonnull
+    private static Map<String, LuceneAnalyzerFactory> initRegistry() {
+        final Map<String, LuceneAnalyzerFactory> registry = new HashMap<>();
+        for (LuceneAnalyzerFactory factory : ServiceLoader.load(LuceneAnalyzerFactory.class)) {
+            final String name = factory.getName();
+            if (registry.containsKey(name)) {
+                if (LOGGER.isWarnEnabled()) {
+                    LOGGER.warn(KeyValueLogMessage.of("duplicate lucene analyzer", LogMessageKeys.ANALYZER_NAME, name));
+                }
+            } else {
+                if (LOGGER.isInfoEnabled()) {
+                    LOGGER.info(KeyValueLogMessage.of("found lucene analyzer", LogMessageKeys.ANALYZER_NAME, name));
+                }
+                registry.put(name, factory);
+            }
+        }
+        return registry;
+    }
+
+    @Nonnull
+    public static LuceneAnalyzerRegistry instance() {
+        return INSTANCE;
+    }
+
+    private LuceneAnalyzerRegistryImpl() {
+        registry = initRegistry();
+    }
+
+    @Nonnull
+    @Override
+    public Map<String, LuceneAnalyzerFactory> getRegistry() {
+        return Collections.unmodifiableMap(registry);
+    }
+
+    @Nonnull
+    @Override
+    public Analyzer getLuceneAnalyzer(@Nonnull Index index) {
+        final String name = index.getOption(IndexOptions.TEXT_ANALYZER_NAME_OPTION);
+        // TODO: Get rid of the condition after OR operator, after having all analyzers registered with this registry
+        if (name == null || !name.equals(NgramAnalyzer.getName())) {
+            return new StandardAnalyzer();
+        } else {
+            LuceneAnalyzerFactory analyzerFactory = registry.get(name);
+            if (analyzerFactory == null) {
+                throw new MetaDataException("unrecognized lucene analyzer for tokenizer", LogMessageKeys.ANALYZER_NAME, name);
+            }
+            return analyzerFactory.getAnalyzer(index);
+        }
+    }
+
+    // Synchronize this method so that we don't need a ConcurrentHashMap but so that
+    // it is still thread safe.
+    @Override
+    @SuppressWarnings("PMD.CompareObjectsWithEquals")
+    public synchronized void register(@Nonnull LuceneAnalyzerFactory analyzerFactory) {
+        final String name = analyzerFactory.getName();
+        LuceneAnalyzerFactory oldFactory = registry.putIfAbsent(name, analyzerFactory);
+        // If there was a factory already registered and the old factory isn't the same as
+        // the new one, throw an error
+        if (oldFactory != null && oldFactory != analyzerFactory) {
```

Comment:
Maybe use `.equals`? I can't actually think of a legit reason to register the same factory twice, but if there were one, I could well imagine that it would include registering an "equivalent" one. Of course, you'd have to override `equals` in that factory class for that to work.

Reply:
Fixed